### PR TITLE
Fixes Django 2.1 breaking changes; QUERY_TERMS & auth_views.login()

### PR DESCRIPTION
**Pull request by xfxf**
Fixes breaking changes in Django 2.1 which is due for release on the 1st August; in ~6 days.
Would recommend a core developer carefully reviews my changes before accepting; if Wagtail 2.2 isn't being released in the next week, may necessitate a 2.1.x minor release ASAP, too (edit: just noticed that setup.py prevents installation with Django 2.1, so may be OK for current Wagtail 2.1.x).

### Fix 1 - QUERY_TERMS

``django.db.models.sql.constants.QUERY_TERMS`` is removed in Django 2.1.
This isn't mentioned in the release notes :(  (I've added a bug there, so it can be added).

This is imported in ``wagtail.contrib.modeladmin.views``.

It was removed on 2017-07-05, link to Django repo:
https://github.com/django/django/commit/244cc401559e924355cf943b6b8e66ccf2f6da3a

It appears the checks for field validity have been removed.  I assume the rationale for this change is if the field is invalid, an Exception will be thrown elsewhere, so it's superfluous having explicit checks which are likely database conditional anyway.

The safest option for a fix appears to be what django-filter has done, which is to add compatibility terms here:
https://github.com/carltongibson/django-filter/pull/921/files

It may be preferential to rewrite lookup_allowed() to not need QUERY_TERMS, but it's probably best a core developer who best understands this code does this.

This PR at least safely prevents an upgrade to Django 2.1 breaking Wagtail; which is likely going to start happening in a week.

### Fix 2 - django.contrib.auth.views methods

``django.contrib.auth.views.login()`` et al were depreciated in 1.11, and removed in 2.1

See the release notes:
https://docs.djangoproject.com/en/2.1/releases/2.1/#features-removed-in-2-1

I've just changed this to LoginView.as_view() instead and moved the template_name into this as a parameter.

Manually tested, both fixes provide a working Wagtail for me (both on the latest RC for Django 2.1 and 2.2-master).  I've updated tox.ini to test for Django 2.1 too, but it's not on pypi so the automated tests won't check against 2.1 until it's released.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -42,7 +42,7 @@ deps =
 
     dj111: Django>=1.11b1,<2.0
     dj111-mssql: django-pyodbc-azure==1.11.0.0
-    dj20: Django>=2.0,<2.1
+    dj20: Django>=2.0,<2.2
```
Comment (auvipy) [inline tox.ini]: instead of changes this you should use dj21: Django>=2.1rc1,<2.2

File: wagtail/contrib/modeladmin/views.py [hunk 2]
```diff
@@ -34,6 +33,18 @@
 from .forms import ParentChooserForm
 
 
+try:
```
Comment (auvipy) [inline wagtail/contrib/modeladmin/views.py]: you might have also look into this way of django-filter adding compatibility for query terms https://github.com/carltongibson/django-filter/pull/921/files

File: tox.ini (hunk 1 above)
Comment (xfxf) [reply tox.ini]: ack; done this.

File: wagtail/contrib/modeladmin/views.py (hunk 2 above)
Comment (xfxf) [reply wagtail/contrib/modeladmin/views.py]: Do you mean https://github.com/carltongibson/django-filter/pull/851 ?

921 is exactly what is implemented here.  It seems the most 'safe' way of solving this for now; note my comments about this in the PR description.

Comment (auvipy) [reply wagtail/contrib/modeladmin/views.py]: got it

**Comment by xfxf** [comment]
tox/travis is failing due to Django 2.1 not being an installable library at the moment; as expected.  That part may not work until 2.1 is actually released.  Should I leave the lines in .travis.yml?

I also logged this with upstream Django (that this removal should at least be in the install docs', and it seems there's a get_lookups() function which can be used instead - see https://github.com/django/django/pull/10227

**Comment by gasman** [comment]
@xfxf Pointing to the git repo in tox.ini will probably do the job (we did something similar in 71ce47bc5aa006f8e23e57f3b29d7f70572c45de):

    dj21: git+https://github.com/django/django.git@2.1rc1#egg=Django

**Comment by carltongibson** [comment]
> It may be preferential to rewrite lookup_allowed() to not need QUERY_TERMS...

I'd recommend going this way. Something like...

```
if len(parts) > 1 and parts[-1] in Field.get_lookups():
```

... should be roughly equivalent. 
(I've only glanced so it may not be exactly what you need!)

**Comment by gasman** [comment]
Thanks @xfxf!
Will bump this to 2.3 as it looks like there are various other fixes required, including supporting the `renderer` argument on form widgets.

**Comment by gasman** [comment]
Merged in a657a75cd7822bd29913f0a48e37eb97ecfd4175, along with further tweaks to the Travis config (upgrading Postgres to a compatible version, dropping the Python 3.4 environment as Django 2.1 no longer supports it, and marking as allow-failures while we work through the remaining fixes). Thanks again!

**Comment by carltongibson** [comment]
@gasman Did you not see my comment? This is the wrong fix. 

Replicating `QUERY_TERMS` in Wagtail is just introducing a technical debt. Again, instead use `get_lookups()`

**Comment by gasman** [comment]
@carltongibson I did see it, but I decided that having a working solution now would be better than waiting for an ideal one. If you put your version into a pull request, I'll be happy to review (and if you can provide a test for the behaviour implemented here - assuming we don't have one already - even better)!

**Pull request closed without merging**